### PR TITLE
feat(promptcompression): pluggable domain-aware NLP compression pipeline

### PR DESCRIPTION
**Pull request by rootfs**
This feat splits the compression into pre-classification and pre-inference so to adopt different compression optimizers.

### CODE REVIEW

**Comment by Xunzhuo** [comment]
Sorry about this. #1553 introduced a large recent refactor, which caused conflicts here. Please rebase onto main and resolve them.

**Review by rootfs — COMMENTED** [review]
## Code Review: PR #1564 — Pluggable Domain-Aware Prompt Compression Pipeline

**Verdict**: `approve`  
This PR cleanly implements a composable NLP compression system with solid architectural separation of UC1 (pre-classification) and UC2 (pre-inference) concerns. The plugin registry design is extensible, test coverage is strong, and integration points are minimal and non-breaking.

---

### Risk

**Low risk; backwards-compatible.**

- **config/model_config_types.go:95–134** — `PromptCompressionConfig` struct extended with optional fields (`InferenceEnabled`, `DefaultPipeline`, `DomainPipelines`). Existing configs continue to work.
- **pkg/extproc/router.go:46–51** — New field `InferenceCompressor` is optional; nil when disabled.
- **pkg/extproc/req_filter_inference_compression.go:107** — Role check includes `"developer"` alongside standard roles. Verify this is supported in the OpenAI API spec you're targeting; if not, remove or gate behind a version flag.
- **Path resolution in router_build.go:149–157** — Relative YAML paths are resolved relative to the router config file's directory. This works but can be fragile if configs are moved. Consider documenting in README the expected directory structure.

---

### Tests

**Strong coverage; no critical gaps.**

- **pipeline_test.go:1–1450** — Comprehensive test suite: YAML parsing, weight normalization, edge cases (empty budget, single sentence, invalid regex). Good.
- **Missing:** Integration tests wiring the full router → decision → pipeline selection path. UC1 and UC2 should both be exercised end-to-end with mock vLLM requests.
- **Missing:** Concurrency test for `PipelineSelector` (it uses `sync.RWMutex`); verify no deadlocks under concurrent `Select()` calls.
- **pipeline_test.go:~250+** — Tests for `CompressMessages` role-aware compression should verify the `SentenceRoles` slice alignment with sentences after splitting. Spot-check one test to confirm indices match.

---

### Style/Hygiene

**Clean; minor observations.**

- **opt_extensions.go:144–152** — Default `defaultPatternBoosts` list is unexported but well-structured. Apply it when YAML has no `patterns` key (line ~235). This implicit fallback is reasonable but worth documenting: "If patterns is empty, built-in Claude Code patterns are used."
- **opt_extensions.go:274–290** — `anyToFloat64` handles many numeric types from YAML unmarshalling; solid defensive coding.
- **req_filter_inference_compression.go:68–75** — JSON marshal/unmarshal on request body is efficient (avoids full message parsing). Good.
- **Comments** are excellent: pipeline.go, README.md, and domain config YAMLs all clearly explain intent and trade-offs.
- **No dead code or commented-out sections** identified.

---

### Size

**Scope is appropriate; PR is a complete feature slice.**

- **3723 additions** — Justified:
  - 5 domain-specific pipeline configs (~77–85 lines each) provide immediate value for UC2.
  - ~476 lines of `pipeline.go` (parsing, execution engine) are necessary.
  - ~335 lines of `opt_extensions.go` (5 optimizer implementations) are standalone but cohesive.
  - ~1450 lines of tests are thorough, not bloat.
- **No arbitrary split needed** — all pieces work together (parsers → registry → pipelines → router integration).
- **Incremental:** UC1 (pre-classification) remains unchanged; UC2 is additive behind a feature flag.

---

### Detailed Observations

| Issue | Location | Severity | Notes |
|-------|----------|----------|-------|
| Role "developer" unsupported? | req_filter_inference_compression.go:107 | ⚠️ **medium** | Verify against OpenAI API. If unsupported, remove to avoid silent failures on non-standard messages. |
| YAML path resolution fragility | router_build.go:149 | ℹ️ **low** | Relative paths are convenient but can break if deployed layout changes. Document expected structure; consider absolute path requirement for production. |
| Concurrency tests for PipelineSelector | pipeline_selector.go:12–30 | ℹ️ **low** | RWMutex is correct, but concurrent stress-test (`Select()` + `Register()` interleaved) would be reassuring. |
| Integration tests missing | — | ℹ️ **low** | Unit tests are solid, but e2e router flow (UC1 + UC2) should be exercised with mock classification results. |
| Default pattern fallback implicit | opt_extensions.go:235 | ℹ️ **info** | When no `patterns` configured, defaults are silently applied. Document this behavior in README or log at load time. |

---

### Recommendations for Merge

1. ✅ **Approve as-is** if the `"developer"` role is confirmed to be valid OpenAI API.
2. **Before merge:** Add one integration test (`extproc_test.go` or similar) that wires a mock request through `modifyRequestBodyForAutoRouting` with `InferenceCompressor` active, verifying a message is actually compressed.
3. **Nice-to-have:** Add a log line at router startup: `"[InferenceCompression] Enabled with N domain pipeline(s), default: <path>"` — already done in router_build.go:126! ✓
4. **Document:** Add the path resolution rules to `pkg/promptcompression/README.md` (suggest relative paths in `config/prompt-compression/`).

File: src/semantic-router/pkg/extproc/req_filter_inference_compression.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+			continue
+		}
+
+		tokens := promptcompression.CountTokensApprox(content)
+		if tokens <= pipeline.MaxTokens {
+			out[i] = raw
+			continue
+		}
+
+		result := promptcompression.CompressWithPipeline(content, pipeline)
```
Comment (rootfs) [inline src/semantic-router/pkg/extproc/req_filter_inference_compression.go:116]: This path compresses each message as flat text, so `role_weight` and `age_decay` cannot work as advertised in the multi-turn presets. If UC2 needs per-turn scoring, thread the conversation roles/ages into the compressor instead of calling `CompressWithPipeline` per message.

File: src/semantic-router/pkg/promptcompression/pipeline.go [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+	if effectiveBudget <= 0 || originalTokens <= effectiveBudget {
+		return Result{
+			Compressed:       text,
+			OriginalTokens:   originalTokens,
+			CompressedTokens: originalTokens,
+			Ratio:            1.0,
+		}
+	}
+
+	sentences := SplitSentences(text)
```
Comment (rootfs) [inline src/semantic-router/pkg/promptcompression/pipeline.go:172]: The single-sentence early return means long one-paragraph inputs, stack traces, and code blocks can bypass compression entirely even when they exceed budget. Consider letting the pipeline run for one-sentence inputs too, or at least gate this on actual token count rather than sentence count.

**Review by rootfs — COMMENTED** [review]
request_changes

The pipeline is a solid direction, but the new UC2 path has a metadata-routing bug and a common no-op case that make the compression behavior unreliable on the prompts this feature targets.

**Risk**
- `src/semantic-router/pkg/promptcompression/opt_extensions.go:29-68`, `src/semantic-router/pkg/promptcompression/optimizer.go:22-31`, `src/semantic-router/pkg/promptcompression/pipeline.go:184-206`: `dedup` can remove/reorder sentences before role metadata is attached, but `OptimizerContext.SentenceRoles` is still indexed by sentence position. That makes `role_weight`/`age_decay` mis-score the remaining sentences once the multi-turn pipeline combines those phases.
- `src/semantic-router/pkg/promptcompression/pipeline.go:172-179`: `CompressWithPipeline` returns unchanged text for any input that splits into one sentence, so long stack traces, file dumps, code blocks, and many log blobs can silently bypass compression even though the new domain configs are optimized for exactly those inputs.
- `src/semantic-router/pkg/extproc/router_build.go:118-158`: enabling `inference_enabled` now makes startup depend on external YAML pipeline files; missing or invalid domain configs become hard failures rather than degraded behavior.

**Tests**
- `src/semantic-router/pkg/promptcompression/pipeline_test.go:13-1450` gives broad unit coverage for the new package.
- Missing coverage is in the actual UC2 integration points: `src/semantic-router/pkg/extproc/router_build.go:118-158` and `src/semantic-router/pkg/extproc/processor_req_body_routing.go:72-90`.
- There’s no test proving `dedup` still preserves correct role/age scoring when combined with `role_weight` or `age_decay`.

**Style/Hygiene**
- The new package is heavily documented, but several comments restate the same pipeline story across YAML, Go docs, and README.
- `opt_extensions.go` and the YAML presets duplicate a lot of pattern/weight prose; a smaller shared reference would be easier to maintain.

**Size**
- The scope is coherent, but it’s large enough that splitting the plugin framework from the router wiring would have reduced review and regression risk.

File: src/semantic-router/pkg/extproc/processor_req_body_routing.go [hunk 3]
```diff
@@ -69,6 +69,26 @@ func (r *OpenAIRouter) modifyRequestBodyForAutoRouting(
...
+	// This runs AFTER serialization so it operates on raw JSON bytes — the same
+	// pattern used by injectMemoryMessages — and BEFORE reasoning/system-prompt
+	// mutations so those additions are never accidentally compressed away.
+	if r.InferenceCompressor != nil {
+		pipeline := r.InferenceCompressor.Select(decisionName)
+		if pipeline.MaxTokens > 0 {
+			compressed, cerr := compressMessagesInBody(modifiedBody, pipeline)
+			if cerr != nil {
+				logging.Warnf("[InferenceCompression] Skipping: %v", cerr)
+			} else if len(compressed) < len(modifiedBody) {
```
Comment (rootfs) [inline src/semantic-router/pkg/extproc/processor_req_body_routing.go:84]: This gate compares JSON byte length, not the compression objective (token reduction). A payload can have fewer tokens but a slightly larger JSON encoding, causing UC2 compression to be dropped silently; decide based on token counts from `compressMessagesInBody` instead.

File: src/semantic-router/pkg/promptcompression/pipeline_selector.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+// NewPipelineSelector builds a PipelineSelector from a SelectorConfig,
+// loading all referenced YAML pipeline files.
+func NewPipelineSelector(cfg SelectorConfig) (*PipelineSelector, error) {
+	s := &PipelineSelector{
+		byDecision: make(map[string]Pipeline, len(cfg.DomainPipelines)),
+	}
+
+	// Load the default pipeline.
+	if cfg.DefaultPipeline == "" {
```
Comment (rootfs) [inline src/semantic-router/pkg/promptcompression/pipeline_selector.go:96]: When `default_pipeline` is empty, `inference_enabled` still succeeds but defaults to `DefaultPipeline(0)` (pass-through). This creates a silent misconfiguration path where inference compression is effectively disabled; return an error when UC2 is enabled without a usable default budget.

File: src/semantic-router/pkg/promptcompression/pipeline_selector.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+// In the router config YAML this appears under prompt_compression:
+//
+//	prompt_compression:
+//	  # Use Case 1 (existing)
+//	  enabled: true
+//	  max_tokens: 512
+//
+//	  # Use Case 2 (new)
+//	  inference_enabled: true
+//	  inference_max_tokens: 4096   # per-pipeline overrides this
```
Comment (rootfs) [inline src/semantic-router/pkg/promptcompression/pipeline_selector.go:67]: This comment references `inference_max_tokens`, but that field is not present in `PromptCompressionConfig`. Please remove or align this example with the actual schema to avoid configuration drift.

**Review by rootfs — COMMENTED** [review]
**request_changes**  
The core architecture is strong, but there are two behavior-level issues that can silently disable or skip inference compression in production paths.

### 2. **Risk**
1. **Compression application is gated on JSON byte size, not compression objective.**  
   Inference compression is only applied when `len(compressed) < len(modifiedBody)` (`src/semantic-router/pkg/extproc/processor_req_body_routing.go:84`), which can reject token-reduced prompts if JSON encoding overhead grows; this can silently negate the feature.

2. **`inference_enabled` can succeed while doing no compression due to implicit zero-budget fallback.**  
   When `default_pipeline` is omitted, selector default becomes `DefaultPipeline(0)` (`src/semantic-router/pkg/promptcompression/pipeline_selector.go:96`, `src/semantic-router/pkg/promptcompression/pipeline_selector.go:99`), i.e., pass-through. With `inference_enabled: true`, this is a high-risk silent misconfiguration unless explicitly validated as error.

### 3. **Tests**
- **Good coverage exists for pipeline mechanics** (parse/registry/weights/optimizers in `src/semantic-router/pkg/promptcompression/pipeline_test.go:13` onward).
- **Missing targeted tests for UC2 routing integration paths**:
  1. No test demonstrated for byte-length gate behavior in request-body mutation flow (`src/semantic-router/pkg/extproc/processor_req_body_routing.go:78`-`src/semantic-router/pkg/extproc/processor_req_body_routing.go:88`).
  2. No test shown that `inference_enabled` + missing `default_pipeline` fails fast or emits deterministic behavior (`src/semantic-router/pkg/extproc/router_build.go:120`-`src/semantic-router/pkg/extproc/router_build.go:128`, `src/semantic-router/pkg/promptcompression/pipeline_selector.go:96`-`src/semantic-router/pkg/promptcompression/pipeline_selector.go:100`).

### 4. **Style/Hygiene**
Comment/doc volume is very high relative to logic changes, and there is some source-of-truth drift risk between examples/comments and config structs (e.g., `inference_max_tokens` appears in selector docs but not in the struct: `src/semantic-router/pkg/promptcompression/pipeline_selector.go:67`).

### 5. **Size**
Scope is large for one PR (new plugin framework, router wiring, configs, demo, docs, and 1450-line test file). This would review/land more safely as:
1. promptcompression engine + registry,  
2. extproc/router UC2 wiring,  
3. domain configs/docs/demo.

**Comment by Xunzhuo** [comment]
this is covered in https://github.com/vllm-project/semantic-router/pull/2004
